### ISSUE TITLE
compile RapidCFD on GPU machine

### ISSUE THREAD

**Issue by irfanms92** [Atizar/RapidCFD-dev#75]
Dear All,
I am new to this. I would like to compile RapidCFD ver 2.3.1 on Ubuntu 16.01 with NVIDIA CUDA toolkit 8.0. Please let me know the procedure to enable 'CUDA' and other tools required to compile the source. 
Thanks in advance.

**Comment by TonkomoLLC** [comment]
Hello,

I think the general instructions on homepage README are correct:

```
Compilation Instructions for Ubuntu 16.04:

ensure CUDA 7.5 is not installed from Ubuntu repsitories
ensure you are using a nVidia driver compatible with CUDA 8 e.g., for the nVidia Tesla K20: http://www.nvidia.com/download/driverResults.aspx/118962/en-us
download CUDA 8.0 here: https://developer.nvidia.com/cuda-80-ga2-download-archive
install CUDA 8.0 per instructions provided at time of download
to compile in parallel, export WM_NCOMPPROCS=10 (as appropriate for your computer)
software should compile without errors. Result: RapidCFD for Ubuntu 16.04
ThirdParty-dev is needed for multiple GPU's.
```

Nonetheless, there are small details missing.

For example, you will need to ensure that the CUDA path is set correctly.  e.g., you need to add the following to your bashrc or equivalent (e.g., setting these parameters before you compile RapidCFD).  Edit the paths as required for your particular system, but generally these should be correct if you did a default install of CUDA 8.

```
export CUDA_HOME=/usr/local/cuda
export PATH=${CUDA_HOME}/bin:${PATH}
export LD_LIBRARY_PATH=${CUDA_HOME}/lib64:$LD_LIBRARY_PATH
```

To be fair, this is part of the step `install CUDA 8.0 per instructions provided at time of download` noted on the main page readme.

To compile RapidCFD,  from the RapidCFD directory, `source etc/bashrc` and then type `./Allwmake`  This is standard for the compilation of all the main OpenFOAM forks.  

Between these additional instructions and what is on the RapidCFD main page readme, hopefully you can compile RapidCFD.

Best regards, and happy RapidCFD foaming! 

Eric

**Comment by ipe-zhangyz** [comment]
Hello, I want to know which version of OpenFOAM is RapidCFD forked from, 2.3.0, 2.3.1 or higher? It seems that there are no tools such as blockMesh in RapidCFD. I used the blockMesh from OpenFOAM 7 to generate a test case, but the icoFoam of RapidCFD failed to read the input file. I also want to know if RapidCFD can only use CUDA 8 and Ubuntu 16.04. I test the icoFoam of RapidCFD  on CUDA 10.1 and CentOS 7.6. After successfully reading the input files generated by OpenFOAM 2.3.0, the icoFoam of RapidCFD crashed, throwing this information:
Courant Number mean: 0 max: 0
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  transform: failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered
Aborted (core dumped)

**Comment by TonkomoLLC** [comment]
I think RapidCFD is forked from 2.3.0.  You are correct, you need blockMesh and other utilities from a full CPU version of OpenFOAM.   I have tested compiling RapidCFD up to CUDA 10.2.  I have not tested CUDA 11.

If you use CUDA 9 or higher there is an error at the end of each execution of a RapidCFD solver that looks like what you described.  Assuming that this is the same error that you are reporting, the cause of this thrust::system::system_error and the way to fix it is discussed [here](https://github.com/Atizar/RapidCFD-dev/issues/59#issuecomment-487136615). Be sure to fully recompile RapidCFD after making any edits to the CUDA header files.

However, as for the crash, I would firstly ensure that the case is appropriately setup for OpenFoam 2.3.0 and not OpenFOAM 7.  If this was already done, then I do not know the solution given the provided information.

Best,
Eric

**Comment by ipe-zhangyz** [comment]
Hi Eric,
Thanks for your reply! I have recompiled the RapidCFD following the instructions you give at https://github.com/Atizar/RapidCFD-dev/issues/57 and then I can run icoFoam without error information. I guess it is the change of sm_30 to sm_60 that works, as the CUDA Capablilty of my GPU is 6.0. I've alse tested some other cases, including the damBreak case from OpenFOAM 2.3.0. In all the tested cases, I find that the GPU version, i.e. RapidCFD is much slower than the OpenFOAM 2.3.0. Is it because the tutorial cases are too small for GPU running? I want to know if you have any cases that RapicCFD is really much faster than OpenFOAM. 

Thanks,
zhangyz

**Comment by TonkomoLLC** [comment]
Hello, Zhangyz,

Glad you are up and running!  I placed some test cases [here](https://github.com/TonkomoLLC/RapidCFD-Tests).  As you noted, one needs a sizable grid to see the speed up with RapidCFD.  But the trick is, especially with older GPU's, is that larger grids require more GPU memory... and you can find yourself limited with one GPU card.  

I generally see 1 K20 GPU = ~8-16 cores of E5-2670 CPU's.  This technology is several years old; I do not have stats for newer technology. 

Good luck with your testing.

Best,

Eric

**Comment by irfanms92** [comment]
RapidCFD is a fork of OpenFOAM version 2.3.1. I have tested RapidCFD with CUDA toolkit 10.2 on Ubuntu 16.04. The issue terminate called after throwing an instance of 'thrust::system::system_error' is observed due to sm_30 set from RapidCFD script. Before compiling source, sm_30 to sm_xx based on the compute capability of your GPU.

**Comment by ipe-zhangyz** [comment]
With a damBreak case of about 2.6 million cells, I see about 10x speed up on one nvidia P100 GPU compared to one Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz CPU.  I failed to run it in parallel mode. With the command `mpirun -np 4 interFoam -parallel -devices "(0 1 2 3)"` I got the error message like this: 
`[0] --> FOAM FATAL ERROR: `
`[0] interFoam: cannot open case directory "/home/zhangyz/RapidCFD/test/damBreakFine3/processor0"
`
What's more, the OpenFOAM 2.3.0 I installed on the same machine can not run in parallel either, throwing  the same error message. 
I'm sure I have write permission in that directory, but I don't know why OpenFOAM and RapidCFD throw that error.  If I mkdir  processor0 in that directory, then the error becomes:
`[0] --> FOAM FATAL ERROR: `
`[0] Cannot find file "points" in directory "polyMesh" in times 0 down to constant
`
Using the same mpirun I can run other programs in parallel on the same machine. 

I'm not familiar with OpenFOAM yet.  Do you have any ideas about this problem?

**Comment by TonkomoLLC** [comment]
Hello, Zhangyz,

The error comes from your not having decomposed the case for multiple GPU cards. This requires the step:

```
decomposePar
```

Before running interFoam.  Since utilities are not included with RapidCFD, you'll need to run `decomposePar` using CPU-based OpenFOAM 2.3.x or whatever version you want.  

Note that if you try to run the case with OpenFOAM 2.3.x, you'll need to execute the solver differently:

```
mpirun -np # interFoam -parallel
```

where `#` is the number of processors, e.g., "4"

With respect to running DamBreak in parallel with RapidCFD, please make sure that you have a directory `ThirdParty-dev` installed.  This topic is discussed [here](https://github.com/Atizar/RapidCFD-dev/issues/2#issuecomment-105454598).   Also, I placed my RapidCFD ThirdParty-dev [here](https://github.com/TonkomoLLC/ThirdParty-dev).  However, I think with recent updates to RapidCFD, the openmpi version that's used is different than the ones in these repositories.  Therefore, some assembly is required. 

- Ensure that in $WM_PROJECT_DIR/etc/config/settings.sh, 

```
export FOAM_MPI=openmpi-4.0.2
```

the value `4.0.2` is set to whatever OpenMPI version you're using.  In these ThirdParty directories that i referred to above, it is not 4.0.2.  

- You may need to change directory (cd) to $WM_PROJET_DIR/src/PStream and type `wclean` to get RapidCFD to recognize the ThirdParty directory, but I am not 100% sure of this comment.  In any case, it does not take long to recompile the PStream directory if I advised you to do an unnecessary step.


I hope I didn't overlook something in the installation of the ThirdParty-dev directory.  It's been a while since I did this installation.

Best regards,

Eric

**Comment by ipe-zhangyz** [comment]
Hi Eric,

Thank you very much for the quick and detailed reply! It's my fault not noticing the decomposePar command in the OpenFOAM user guide. All other parts of my installation are OK. Now I can run both OpenFOAM 2.3.0 and RapidCFD in parallel successfully. It seems that the scalability of RapidCFD is not as good as OpenFOAM. With the same damBreak case, I see nearly 4x speed up when running on 4 CPUs compared to one CPU running, but only a little more than 2x speed up of 4 GPUs running compared to one GPU running.  Maybe I can improve it by using some binding strategies, but I'm not sure if they have effects. Anyway, the GPU running is much faster than the CPU running when the case is large enough, so it's good to have this GPU version of OpenFOAM. Thanks for your contribution!

There's another thing that I'm wondering.  It seems that RapidCFD is still in maintain right? As the OpenFOAM develops to version 7, why is RapidCFD still based on OpenFOAM 2.3.x? Is it very difficult or time comsuming for RapidCFD to follow the evolution of OpenFOAM? On the other hand, why doesn't the owner of OpenFOAM consider adding the GPU implementation to their official release version, as there are so many applications having supported GPU acceleration now? Maybe these questions are a little far from the topic of this issue. I'm just curious...

Thanks,
zhangyz

**Comment by TonkomoLLC** [comment]
Hi, Zhangyz,

I'm glad you got everything working.  Thanks for the scaling report.  I've seen various reports on the scaling efficiency of RapidCFD.  There are a lot of details with scaling.  For example, if there are a fixed number of cells...  If you divide that fixed number of cells with more GPU's, I think the scaling efficiency can go down like you saw (i.e., the strong scaling efficiency of RapidCFD).   However, if you and an optimum number of cells per GPU for a given problem, and you scale up such that you fix the number of cells per GPU at the optimum, then the scaling efficiency can be quite high (here I refer to the weak scaling efficiency, where we have a fixed problem size per processor).  

As to why RapidCFD is still at OpenFOAM 2.3?  I cannot speak for @daniel-jasinski , but as for myself, I simply do not have the time to invest in upgrading RapidCFD to OpenFOAM 7. Most of my day-to-day work involves reactive CFD, which is not applicable for RapidCFD (and I suspect for GPGPU computing, in general).  So it is a little bit difficult for me to get the time to massively improve the code.  I try to show my appreciation for the work done by sharing my learning with others in the issues board on this GitHub repo.

If you are interested in improving RapidCFD, I am sure @daniel-jasinski 's [words](https://sim-flow.com/rapid-cfd-gpu/?fbclid=IwAR2RgfMZS-ujYMzFUs1Duo7d3HJLfaYORB_WwuP9wh6QinPyGITI-qUj5Yc) are still true:

```
With the release of RapidCFD, we want to give everyone an opportunity to contribute. If you have the brains and skills, we’d love for you to knock out any RapidCFD issues.
```

I hope that you have continued success with RapidCFD and your projects.

Best regards,

Eric

**Comment by ipe-zhangyz** [comment]
Hi Eric,

It's very kind of you to always give quick and detailed reply! I've learned a lot from you. Thank you very much!

Sincerely,
zhangyz
